File: tests/correct/data-structures/treiber-stack-dynamic/my_stack.c
#include <stdlib.h>
#include <pthread.h>
#include <stdatomic.h>
#include <genmc.h>

#include "my_stack.h"

#ifdef MAKE_ALL_SC
# define release memory_order_seq_cst
# define acquire memory_order_seq_cst
# define relaxed memory_order_seq_cst
#else
# define release memory_order_release
# define acquire memory_order_acquire
# define relaxed memory_order_relaxed
#endif

#ifndef MAX_THREADS
# define MAX_THREADS 32
#endif

#define POISON_IDX 0x666

static node_t *new_node()
{
	return malloc(sizeof(node_t));
}

static void reclaim(node_t *node)
{
	__VERIFIER_hp_retire(node);
}

void init_stack(mystack_t *s, int num_threads)
{
	/* initialize stack */
	atomic_init(&s->top, 0);
}

void clear_stack(mystack_t *s, int num_threads)
{
	node_t *next;
	while (s->top != 0) {
		next = s->top->next;
		free(s->top);
		s->top = next;
	}
}

void push(mystack_t *s, unsigned int val)
{
	node_t *node = new_node();
	node->value = val;

	while (true) {
		node_t *top = atomic_load_explicit(&s->top, acquire);
		atomic_store_explicit(&node->next, top, relaxed);
		if (atomic_compare_exchange_strong_explicit(&s->top, &top, node, release, relaxed))
			break;
	}
}

unsigned int pop(mystack_t *s)
{
	node_t *top;
	int val;

	__VERIFIER_hp_t *hp = __VERIFIER_hp_alloc();
	while (true) {
		top = __VERIFIER_hp_protect(hp, &s->top);
		if (top == NULL) {
			__VERIFIER_hp_free(hp);
			return 0;
		}

		node_t *next = atomic_load_explicit(&top->next, relaxed);
		if(atomic_compare_exchange_strong_explicit(&s->top, &top, next, release, relaxed))
			break;
	}
	val = top->value;
	/* Reclaim the used slot */
	reclaim(top);
	__VERIFIER_hp_free(hp);
	return val;
}

File: lli/runtime-include/rust/rust/tests/correct/data-structures/treiber-stack-dynamic/my_stack.rs
///!Simple Rust-Implementation of Treiber-Stack tests given in treiber-stack-dynamic for C.
use std::{mem::MaybeUninit, ptr, sync::atomic::{AtomicPtr, Ordering}};
use std::{alloc::{alloc, Layout}, ffi::c_void, genmc::{__VERIFIER_hp_alloc, __VERIFIER_hp_protect, __VERIFIER_hp_retire, __VERIFIER_hp_free, __VERIFIER_hp_t}};
use std::sync::atomic::Ordering::*;

pub const MAX_NODES : usize = 0xff;
pub const MAX_THREADS: usize = 32;

#[repr(C)]
pub struct Node {
    pub value: u32,
    pub next: AtomicPtr<Node>
}

#[repr(C)]
pub struct MyStack {
    pub top: AtomicPtr<Node>,
    //nodes: MaybeUninit<[Node; MAX_NODES + 1]> => not used..
}

impl Node {
    pub fn new() -> *mut Node {
        let ptr = unsafe { alloc(Layout::new::<Node>()) };
        ptr as *mut Node
    }

    pub fn reclaim(node: *mut Node) {
        __VERIFIER_hp_retire(node);
    }
}

impl MyStack {
    pub const fn init_stack() -> MyStack {
        //let ns: [MaybeUninit<Node>; MAX_NODES + 1] = unsafe { MaybeUninit::uninit().assume_init() }; //Important!
        let m = MyStack {
            top: AtomicPtr::new(ptr::null_mut()),
            //nodes: ns
        };
        //m.top.store(ptr::null_mut(), SeqCst);

        m
    }


    pub fn push(&self, val: u32) {
        let node: *mut Node = Node::new();
        unsafe { (*node).value = val };

        loop {
            let top: *mut Node = self.top.load(Acquire);
            unsafe { (*node).next.store(top, Relaxed) };
            if self.top.compare_exchange(top, node, Release, Relaxed).is_ok() {
                break;
            }
        }
    }

    pub fn pop(&self) -> u32 {
        let mut top_: *mut Node = ptr::null_mut();
        let mut val: u32 = 0;

        let hp: *mut __VERIFIER_hp_t = __VERIFIER_hp_alloc();
        loop {
            top_ = __VERIFIER_hp_protect(hp, &(self.top));
            if top_.is_null() {
                __VERIFIER_hp_free(hp);
                return 0;
            }

            let next: *mut Node = unsafe { (*top_).next.load(Ordering::Relaxed) };
            if self.top.compare_exchange(top_, next, Ordering::Release, Ordering::Relaxed).is_ok() {
                break;
            }
        }
        val = unsafe { (*top_).value } ;
        /* Reclaim the used slot */
        Node::reclaim(top_);
        __VERIFIER_hp_free(hp);
        val
    }
}